src/cpp/shared/town/town.cpp

```diff
@@ -606,7 +726,617 @@ char * GetDwellingName(int faction, int dwellingIndex) {
+    GUISetImgIdx(casWin, i + 700, castleSlotsUse[i]);
+
+  sprintf(gText, "cstl%s.icn", cHeroTypeShortName[this->castle->factionID]);
+  for(int i = 0; i < NUM_NON_DWELLING_BUILDINGS; ++i) {
+    GUISetIcon(casWin, i + 700, gText);
+  }
+
+  for(int i = 0; i < NUM_NON_DWELLING_BUILDINGS; ++i) {
+    if(castleSlotsUse[i]) {
+      GUISetText(casWin, i + 600, GetBuildingName(this->castle->factionID, castleSlotsUse[i]));
+    } else {
+      int mageGuildLevel = this->castle->mageGuildLevel + 1;
+      if(mageGuildLevel >= 5)
+        mageGuildLevel = 5;
+      GUISetText(casWin, i + 600, "Mage Guild, Level " + std::to_string(mageGuildLevel));
+    }    
+  }
+
+  int imgIdx;
+  for(int i = 0; i < NUM_NON_DWELLING_BUILDINGS; ++i) {
+    imgIdx = -1;
+    if((1 << castleSlotsUse[i]) & this->castle->buildingsBuiltFlags
+      && (castleSlotsUse[i] || this->castle->mageGuildLevel == 5)) {
+      imgIdx = 11;
+    } else {
+      if((1 << castleSlotsUse[i]) & this->field_156) {
+        if(!((1 << castleSlotsUse[i]) & this->field_152))
+          imgIdx = 13;
+      } else {
+        imgIdx = 12;
+      }
+    }
+    if(imgIdx == -1)
+      GUIRemoveFlag(casWin, i + 800, 4);
+    else {
+      GUIAddFlag(casWin, i + 800, 4);
+      GUISetImgIdx(casWin, i + 800, imgIdx);
+    }
+    if(imgIdx == 11)
+      GUIRemoveFlag(casWin, i + 400, 4);
+    else {
+      GUIAddFlag(casWin, i + 400, 4);
+      if(imgIdx == -1)
+        GUISetImgIdx(casWin, i + 400, 1);
+      else
+        GUISetImgIdx(casWin, i + 400, 2);
+    }
+  }
+
+  int hasCaptain = this->castle->buildingsBuiltFlags & (1 << BUILDING_CAPTAIN);
+  if(hasCaptain) {
+    GUIRemoveFlag(casWin, 1101, 6);
+    GUISetImgIdx(casWin, 1100, 1);
+    GUIAddFlag(casWin, 1106, 4);
+    GUISetImgIdx(casWin, 1106, gpCurPlayer->color);
+  } else {  
+    GUIAddFlag(casWin, 1101, 6);
+    GUISetImgIdx(casWin, 1100, 0);
+    GUIRemoveFlag(casWin, 1106, 4);
+  }
+
+  sprintf(gText, "CSTLCAP%c.ICN", cHeroTypeInitial[this->castle->factionID]);
+  GUISetIcon(casWin, 1100, gText);
+
+  imgIdx = -1;
+
+  if(hasCaptain) {
+    char a2[100];
+    sprintf(gText, "");
+    for(int i = 0; i < 4; ++i) {
+      sprintf(a2, "%s\n", gStatNames[i]);
+      strcat(gText, a2);
+    }
+    GUISetText(casWin, 1104, gText);
+
+    sprintf(gText, "");
+    for(int i = 0; i < 4; ++i) {
+      sprintf(a2, "%d\n", captainStats[this->castle->factionID][i]);
+      strcat(gText, a2);
+    }
+    GUISetText(casWin, 1105, gText);
+
+    if(this->castle->field_38) {
+      GUIRemoveFlag(casWin, 213, 4);
+      GUIAddFlag(casWin, 215, 4);
+    } else {
+      GUIAddFlag(casWin, 213, 4);
+      GUIRemoveFlag(casWin, 215, 4);
+    }
+  } else {
+    if(CanBuild(this->castle, BUILDING_CAPTAIN)) {
+      if(!CanBuy(this->castle, BUILDING_CAPTAIN))
+        imgIdx = 13;
+    } else {
+      imgIdx = 12;
+    }
+    if(CanBuild(this->castle, BUILDING_CAPTAIN))
+      this->field_156 |= 0x8000u;
```

Comment:
It seems field_156 and field 152 are the CanBuild and CanBuy fields, respectively. If you can confirm, then rename and fix the bit shifts.